pkg/generic/descriptor/http.go

```diff
@@ -32,25 +35,105 @@ const (
 	MIMEApplicationProtobuf = "application/x-protobuf"
 )
 
+var (
+	_ dhttp.RequestGetter  = &HTTPRequest{}
+	_ dhttp.ResponseSetter = &HTTPResponse{}
+)
+
 // HTTPRequest ...
 type HTTPRequest struct {
-	Header      http.Header
-	Query       url.Values
-	Cookies     Cookies
-	Method      string
-	Host        string
-	Path        string
 	Params      *Params // path params
+	Request     *http.Request
 	RawBody     []byte
 	Body        map[string]interface{}
 	GeneralBody interface{} // body of other representation, used with ContentType
 	ContentType MIMEType
 }
 
+// GetHeader implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetHeader(key string) string {
+	return req.Request.Header.Get(key)
+}
+
+// GetCookie implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetCookie(key string) string {
+	if c, err := req.Request.Cookie(key); err == nil {
+		return c.Value
+	}
+	return ""
+}
+
+// GetQuery implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetQuery(key string) string {
+	return req.Request.URL.Query().Get(key)
+}
+
+// GetBody implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetBody() []byte {
+	return req.RawBody
+}
+
+// GetMethod implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetMethod() string {
+	return req.Request.Method
+}
+
+// GetPath implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetPath() string {
+	return req.Request.URL.Path
+}
+
+// GetHost implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetHost() string {
+	return req.Request.Host
+}
+
+// GetParam implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetParam(key string) string {
+	return req.Params.ByName(key)
+}
+
+// GetMapBody implements http.RequestGetter of dynamicgo
+func (req *HTTPRequest) GetMapBody(key string) string {
+	if len(req.RawBody) == 0 {
+		return ""
+	}
+
+	body := req.RawBody
+	s := utils.SliceByteToString(body)
+	node := ast.NewRaw(s)
```

Comment:
@jayantxie 
It's possible to access the same key repeatedly in theory, but I don't know whether is should be cached or not, because if there is little repeat, it will be slower.

Reply:
ast.NewRaw has nothing to do with `the same key`, right?

Reply:
@jayantxie Ah I think I got what you mean. Yes we can cache node itself, thank you.

Reply:
@jayantxie Do you think it is ok to add `MapBody` for caching to HTTPRequest?

Reply:
fixed